### PR TITLE
Nova: workloads adoption, mariadb pre-/post-checks

### PR DESCRIPTION
**Pull request by bogdando**
* MariaDB pre-/post-checks:

Get services topology specific configuration in
pull_openstack_configuration. Add missing role for that as well.

Split maridb source vs podified env vars shell headers to use it
in corresponding places.

Update and fix the composition of the services pre-check list to
execute it before stopping services.

Update and fix the composition of the list of the services to be
stopped (cannot pull data from stopped services).

Verify no dataplane disruptions during the FFU/adoption process.

Verify Nova services still control pre-created VM workload after
FFU/adotpion is done.

Replace oc run with oc exec in post checks. Fix 'oc rsh' which breaks
expected formatting for 'mysql -rs', when it is called from console vs
called by ansible shell module. Use instead 'oc exec' to keep post
checks working the same way in docs and executed via ansible tests.

* Pre-launch test VM instance in dev env role.

Add a test role for development environment and mention it
in the docs TOC as well.

Defer pingtest until the network adoption done.
Defer cinder volumes until workloads with volumes adoption done.
Defer Nova RBD backend until workloads adoption from ceph to ceph done.

Use a copy of the install_yaml shell script to bootstrap a VM.
We cannot use it via intall_yamls devsetup make target because:
* we need to wrap it with shell/ocp/etc headers
* we need to use bash aliases ref to openstack commands instead of
  direct openstack CLI syntax
* we should avoid SSH-ing to the source cloud VMs (standalone/computes)
  during EDPM adoption
* install_yaml is going to be deprecated anyway
* it's FIP ping test doesn't work and depends on the network adoption
  done
* it's cinder volume/backup/snapshot/attach test doesn't work yet,
  which is OK for this simpe workload adoption case. 

Jira: [OSPRH-3123](https://issues.redhat.com/browse/OSPRH-3123)

### CODE REVIEW

**Comment by bogdando** [comment]
recheck no logs

**Comment by gibizer** [comment]
```
[mariadb_copy : test connection to podified DBs (show databases)]
...
COLLATION=utf8_general_ci\n/bin/bash: line 17: MARIADB_IMAGE: unbound variable", "stderr_lines":
```

**Comment by gibizer** [comment]
Current adoption test failure:
```
TASK [ovn_adoption : start an adoption helper pod] *****************************
...
=10.217.5.20", "+ PODIFIED_OVSDB_SB_IP=10.217.4.161", "+ OVSDB_IMAGE=quay.io/podified-antelope-centos9/openstack-ovn-base:current-podified", "+ SOURCE_OVSDB_IP=192.168.122.100", "+ CONTROLLER1_SSH='ssh -i /home/zuul/src/github.com/openstack-k8s-operators/install_yamls/out/edpm/ansibleee-ssh-key-id_rsa root@192.168.122.100'", "+ CONTROLLER2_SSH=:", "+ CONTROLLER3_SSH=:", "+ oc delete pod ovn-data --ignore-not-found=true", "+ oc delete pvc ovn-data --ignore-not-found=true", "+ oc apply -f -", "Error from server (BadRequest): error when creating \"STDIN\": Pod in version \"v1\" cannot be handled as a Pod: json: cannot unmarshal object into Go struct field PodSpec.spec.volumes of type []v1.Volume"], "stdout": "persistentvolumeclaim/ovn-data created", "stdout_lines": ["persistentvolumeclaim/ovn-data created"]}
```

**Comment by bogdando** [comment]
> Current adoption test failure:
> 
> ```
> TASK [ovn_adoption : start an adoption helper pod] *****************************
> ...
> =10.217.5.20", "+ PODIFIED_OVSDB_SB_IP=10.217.4.161", "+ OVSDB_IMAGE=quay.io/podified-antelope-centos9/openstack-ovn-base:current-podified", "+ SOURCE_OVSDB_IP=192.168.122.100", "+ CONTROLLER1_SSH='ssh -i /home/zuul/src/github.com/openstack-k8s-operators/install_yamls/out/edpm/ansibleee-ssh-key-id_rsa root@192.168.122.100'", "+ CONTROLLER2_SSH=:", "+ CONTROLLER3_SSH=:", "+ oc delete pod ovn-data --ignore-not-found=true", "+ oc delete pvc ovn-data --ignore-not-found=true", "+ oc apply -f -", "Error from server (BadRequest): error when creating \"STDIN\": Pod in version \"v1\" cannot be handled as a Pod: json: cannot unmarshal object into Go struct field PodSpec.spec.volumes of type []v1.Volume"], "stdout": "persistentvolumeclaim/ovn-data created", "stdout_lines": ["persistentvolumeclaim/ovn-data created"]}
> ```

I didn't plan to push unrelated commits, gonna fix that

File: tests/vars.sample.yaml
Comment (bogdando) [inline tests/vars.sample.yaml:7]: there will be a dependency patch for rdo-jobs to enable that for CI

**Comment by danpawlik** [comment]
recheck

**Comment by bogdando** [comment]
It seems that the pingtest have never worked for pre-created VM in the adoption dev env setup.

Just checked, that ``ping -c4 192.168.122.20`` doesn't work on standalone VM on my env neither. @jistr have you ever checked connectivity to pre-created VM?
I think I'll just remove the connectivity check from the post-checks PR, until resolved
relying on openstack server --diagnostics should be fine

**Comment by bogdando** [comment]
recheck no logs

**Comment by cescgina** [comment]
@bogdando please hold the rechecks for a while, there is a quay update going on and the jobs are failing to login there from the looks of it

**Comment by bogdando** [comment]
fails on the last check to verify if Nova services can start the existing test VM instance,
trying to reproduce on my env

**Comment by bogdando** [comment]
I have reproduced this, the openstack server start command no longer brings the shutoff VM back. That could be a regression after libvirt decontainerization, it used to work earlier for my testing.

**Comment by gibizer** [comment]
> I have reproduced this, the openstack server start command no longer brings the shutoff VM back. That could be a regression after libvirt decontainerization, it used to work earlier for my testing.
> 
> nova compute log (my local env)
> 
> ```
> 2023-12-13 16:53:17.292 2 ERROR oslo_messaging.rpc.server     raise libvirtError('virDomainCreateWithFlags() failed')
> 2023-12-13 16:53:17.292 2 ERROR oslo_messaging.rpc.server libvirt.libvirtError: Failed to connect socket to '/run/libvirt/virtlogd-sock': Connection refused
> ```
> 
> that is because virtlogd has failed to start
> 
> ```
> ...Can't load config file: Failed to open file '/etc/libvirt/virtlogd.conf': Permission denied: /etc/libvirt/virtlogd.conf
> virtlogd.service: Main process exited, code=exited, status=1/FAILURE
> ```
> 
> it is owned by root:root, but it should be root:libvirt maybe as the libvirt group includes the kolla nova user id `1406941 -rw-r-----. 1 root root 106 Dec 13 16:39 /etc/libvirt/virtlogd.conf`

This is due to having an old openstack-selinux package in the EDPM node that does not include the fix https://github.com/redhat-openstack/openstack-selinux/pull/106 . The package `0.8.37-0.20231201082840.165e251.el9` has the fix.

**Comment by bogdando** [comment]
> This is due to having an old openstack-selinux package in the EDPM node that does not include the fix [redhat-openstack/openstack-selinux#106](https://github.com/redhat-openstack/openstack-selinux/pull/106) . The package `0.8.37-0.20231201082840.165e251.el9` has the fix.

I'll try that version. There is still a disconnect with 0640 vs how we want it in EDPM (0644)

**Comment by bogdando** [comment]
> The package `0.8.37-0.20231201082840.165e251.el9` has the fix.

Actually I don't know where to get that package to unblock CI
@marios any hints?

**Comment by gibizer** [comment]
> > The package `0.8.37-0.20231201082840.165e251.el9` has the fix.
> 
> Actually I don't know where to get that package to unblock CI @marios any hints?

when I deploy EDPM nodes with make edpm_compute and then deploy with make edpm_deploy I get that package installed

**Comment by gibizer** [comment]
[root@edpm-compute-0 ~]# dnf info openstack-selinux
Last metadata expiration check: 0:21:24 ago on Thu 14 Dec 2023 11:21:34 AM UTC.
Installed Packages
Name         : openstack-selinux
Version      : 0.8.37
Release      : 0.20231201082840.165e251.el9
Architecture : noarch
Size         : 305 k
Source       : openstack-selinux-0.8.37-0.20231201082840.165e251.el9.src.rpm
Repository   : @System
From repo    : delorean-component-common
Summary      : SELinux Policies for OpenStack
URL          : https://github.com/redhat-openstack/openstack-selinux
License      : GPLv2
Description  : SELinux policy modules for use with OpenStack

Available Packages
Name         : openstack-selinux
Version      : 0.8.37
Release      : 0.20231201082840.165e251.el9
Architecture : src
Size         : 61 k
Source       : None
Repository   : delorean-component-common
Summary      : SELinux Policies for OpenStack
URL          : https://github.com/redhat-openstack/openstack-selinux
License      : GPLv2
Description  : SELinux policy modules for use with OpenStack

**Comment by marios** [comment]
> > > The package `0.8.37-0.20231201082840.165e251.el9` has the fix.
> > 
> > 
> > Actually I don't know where to get that package to unblock CI @marios any hints?
> 
> when I create EDPM nodes with make edpm_compute and then deploy with make edpm_deploy I get that package installed

o/ folks replying so it doesn't look like I ignored this ;)

as discussed in slack the make edpm_compute is creating a centos vm. the current job is using a rhel node so we have to work within the subscription here.

we might see if there is a next gen repo available for us to use with this subscription.

**Comment by SeanMooney** [comment]
we are not ment to be useing any rhel node on the github ci so the tripleo standalone node should be deploying centos not rhel and using rdo.

osp adoption should be tested downstream only.

this should be addressed by  https://github.com/openstack-k8s-operators/edpm-ansible/pull/518
i belveive assuming the repos are set up 

there is also a related short term fix for a downstream cix https://github.com/openstack-k8s-operators/ci-framework/pull/935

in general, we need to confirm that repo_setup is executed or for wonstream subscription manager is run to update the repos to the 18/antelop content then bootstracp runs after that and will do a dnf update providing the new selinux package.

**Comment by bogdando** [comment]
Added missing dependencies which should address the issue

**Comment by bogdando** [comment]
recheck

**Comment by bogdando** [comment]
recheck

**Comment by bogdando** [comment]
recheck

**Comment by bogdando** [comment]
recheck

**Comment by bogdando** [comment]
Info from the held node: 
```
openstack-selinux.noarch                                      0.8.37-1.20221215081212.34c2ecc.el9ost                                       @openstack-17.1-for-rhel-9-x86_64-rpms
```
so this remains blocked until we figure out how to set up osp18 repos

File: docs/README.md
Comment (odyssey4me) [inline docs/README.md:10]: Spelling error

File: docs/openstack/stop_openstack_services.md
Comment (bogdando) [inline docs/openstack/stop_openstack_services.md:82]: actually, this is too early to become a point of no return. I should move that step before starting EDPM adoption, which is an agreed point of no return instead.

Comment (bogdando) [reply docs/openstack/stop_openstack_services.md]: done in the roles split commit

**Comment by bogdando** [comment]
it seems it's time to factor out the "Pre-launch test VM instance in dev env role." into a different PR.
The top commit for galera db extra checks, and VIP usage will be also factored out

**Comment by bogdando** [comment]
I'm puzzled as these workload control checks work on my env:
```
TASK [dataplane_adoption : verify if Nova services can start the existing test VM instance] ***
FAILED - RETRYING: [localhost]: verify if Nova services can start the existing test VM instance (60 retries left).
changed: [localhost] => {"attempts": 2, "changed": true, "cmd": "set -euxo pipefail

alias openstack=\"oc exec -t openstackclient -- openstack\"
FIP=192.168.122.20

${BASH_ALIASES[openstack]} server list | grep -qF '| test | SHUTOFF |' && ${BASH_ALIASES[openstack]} server start test
${BASH_ALIASES[openstack]} server list | grep -F '| test | ACTIVE |'
${BASH_ALIASES[openstack]} server --os-compute-api-version 2.48 show --diagnostics test --fit-width -f json | jq -r '.state' | grep running
",
"delta": "0:00:14.727178",
"end": "2024-01-24 15:02:01.778737",
"rc": 0, "start": "2024-01-24 15:01:47.051559",
"stderr": "+ alias 'openstack=oc exec -t openstackclient -- openstack'
+ FIP=192.168.122.20
+ grep -qF '| test | SHUTOFF |'
+ oc exec -t openstackclient -- openstack server list
+ oc exec -t openstackclient -- openstack server list
+ grep -F '| test | ACTIVE |'
+ oc exec -t openstackclient -- openstack server --os-compute-api-version 2.48 show --diagnostics test --fit-width -f json
+ jq -r .state
+ grep running",
"stderr_lines": ["+ alias 'openstack=oc exec -t openstackclient -- openstack'",
"+ FIP=192.168.122.20",
"+ grep -qF '| test | SHUTOFF |'",
"+ oc exec -t openstackclient -- openstack server list",
"+ oc exec -t openstackclient -- openstack server list",
"+ grep -F '| test | ACTIVE |'",
"+ oc exec -t openstackclient -- openstack server --os-compute-api-version 2.48 show --diagnostics test --fit-width -f json",
"+ jq -r .state",
"+ grep running"], "stdout": "| 727300c8-7f9a-4d62-957f-003668583519 | test | ACTIVE | private=192.168.0.21, 192.168.122.20 | cirros | m1.small |
running",
"stdout_lines": ["| 727300c8-7f9a-4d62-957f-003668583519 | test | ACTIVE | private=192.168.0.21, 192.168.122.20 | cirros | m1.small |",
"running"]}
```

and fail in CI during the start back task...

**Comment by cescgina** [comment]
recheck

**Comment by bogdando** [comment]
> and fail in CI during the start back task...
> 
> ```
> ERROR oslo_messaging.rpc.server nova.exception.DiskNotFound: No disk at /var/lib/nova/instances/a53cc8a6-da5e-416e-9657-8c75653fd6db/disk
> ```

On auto-held node I have caught something else this time
```
2024-01-25 17:25:46.021 2 ERROR oslo_service.periodic_task oslo_messaging.exceptions.MessagingTimeout: Timed out waiting for a reply to message ID d532835d37554092bb9dc36666cf6f51
```

**Comment by bogdando** [comment]
recheck no luck

**Comment by cescgina** [comment]
recheck

**Comment by bogdando** [comment]
the prev job failure was unrelated " Invalid volume: Volume 00f833f9-8f94-4041-908e-ff1dbf3bf1f3 status must be available, but current status is: backing-up. (HTTP 400) (Request-ID: req-38853a0c-a3f9-47b1-86e0-3efb
d41c21f2)", "

**Comment by bogdando** [comment]
noted about it's cinder volume/backup/snapshot/attach test doesn't work yet,
which is OK for this simpe workload adoption case.

**Comment by fultonj** [comment]
> the [prev job](https://logserver.rdoproject.org/93/193/1e1fd75f6cbaf231ffc93eeeca229e653efa6690/github-check/data-plane-adoption-osp-17-to-extracted-crc/790c1f8/controller/data-plane-adoption-tests-repo/data-plane-adoption/tests/logs/test_with_ceph_out_2024-01-26T09:48:53EST.log) failure was unrelated " Invalid volume: Volume 00f833f9-8f94-4041-908e-ff1dbf3bf1f3 status must be available, but current status is: backing-up. (HTTP 400) (Request-ID: req-38853a0c-a3f9-47b1-86e0-3efb d41c21f2)", "
> 
> created https://issues.redhat.com/browse/OSPRH-4217

@bogdando you aksed me to look at this but would you please give me more context on what is happening here and what you need for me?

Does the following describe it?

- You have a workload on an overcloud deployed by tripleo which is using ceph RBD for glance, cinder, nova
- This job runs an adoption to move that overcloud to one manged by k8s-operators and you're testing if the workload survives
- the results indicate that part of the workload, a cinder volume, has a status of `backing up` instead of `active`

Do you have any idea why the volume is in a state of backing up?

**Comment by bogdando** [comment]
This tries to adopt workload form ceph to local storage which wont work.

**Comment by bogdando** [comment]
> Does the following describe it?
> 
> * You have a workload on an overcloud deployed by tripleo which is using ceph RBD for glance, cinder, nova

yes. I need some help with configuring it the same way post-adoption , for EDPM, in order to adopt that workload.

> * This job runs an adoption to move that overcloud to one manged by k8s-operators and you're testing if the workload survives

correct - either it survives shutdown and power on actions

> * the results indicate that part of the workload, a cinder volume, has a status of `backing up` instead of `active`

that happens during an initial bootstrap of a test VM in the source cloud, we can omit this for the time being

**Comment by bogdando** [comment]
recheck

**Comment by fultonj** [comment]
> This tries to adopt workload form ceph to local storage which wont work.

Right. That should not work. Initially deploying with nova on local storage as per the following should work

https://github.com/openstack-k8s-operators/install_yamls/pull/712

**Comment by fultonj** [comment]
> > Does the following describe it?
> > 
> > * You have a workload on an overcloud deployed by tripleo which is using ceph RBD for glance, cinder, nova
> 
> yes. I need some help with configuring it the same way post-adoption , for EDPM, in order to adopt that workload. Ideally, I'd defer ceph to ceph adoption of workloads for now, leaving it for a future work, after this simplistic case is done. But I'm not sure how should we modify the install_yaml's make standalone target to configure ceph and local files backend for Nova. Then we need to change the test VM creation script to use local backend instead @SeanMooney FYI.

So we'll initially configure Nova with local disk, not the Ceph VMs pool.

https://github.com/openstack-k8s-operators/install_yamls/pull/712#pullrequestreview-1851510722

> > * This job runs an adoption to move that overcloud to one manged by k8s-operators and you're testing if the workload survives
> 
> correct - either it survives shutdown and power on actions
> 
> > * the results indicate that part of the workload, a cinder volume, has a status of `backing up` instead of `active`
> 
> that happens during an initial bootstrap of a test VM in the source cloud, we can omit this for the time being

Why would something like this:
```
 openstack server create --flavor $FLAV --volume $VOL_ID --nic net-id=$NET my-pet-vm
```
result in something like this?
```
 openstack volume backup create $VOL_ID
```
I'm not sure this theory on what's happening is correct. 

But as per:

  https://github.com/openstack-k8s-operators/install_yamls/pull/712#pullrequestreview-1851510722
  
We're not going to find out in this PR and we'll investigate what's really going on later via https://issues.redhat.com/browse/OSPRH-4287

**Comment by bogdando** [comment]
recheck

**Comment by bogdando** [comment]
recheck

**Comment by bogdando** [comment]
this fails differently now, and very early 'Cloud standalone was not found', for TASK [development_environment : pre-launch test VM instance]

**Comment by bogdando** [comment]
> Why would something like this:
> 
> ```
>  openstack server create --flavor $FLAV --volume $VOL_ID --nic net-id=$NET my-pet-vm
> ```
> 
> result in something like this?
> 
> ```
>  openstack volume backup create $VOL_ID
> ```

this is  how we are used to do that in the adoption docs historically https://github.com/openstack-k8s-operators/data-plane-adoption/blob/main/docs_dev/assemblies/development_environment.adoc?plain=1#L283

let's leave it out of this PR scope for now. agreed

**Comment by fultonj** [comment]
> > Why would something like this:
> > ```
> >  openstack server create --flavor $FLAV --volume $VOL_ID --nic net-id=$NET my-pet-vm
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > result in something like this?
> > ```
> >  openstack volume backup create $VOL_ID
> > ```
> 
> this is how we are used to do that in the adoption docs historically https://github.com/openstack-k8s-operators/data-plane-adoption/blob/main/docs_dev/assemblies/development_environment.adoc?plain=1#L283 after https://github.com/openstack-k8s-operators/data-plane-adoption/blob/main/docs_dev/assemblies/development_environment.adoc?plain=1#L267

Ah ha! Then, one solution is for the job wait for the backup to complete.

> let's leave it out of this PR scope for now. agreed

ok.

**Comment by bogdando** [comment]
recheck deps changed

**Comment by bogdando** [comment]
recheck deps changed

**Comment by bogdando** [comment]
recheck error: unable to default to a user name: the server is currently unable to handle the request (get users.user.openshift.io ~)

**Comment by bogdando** [comment]
This iteration of workloads adoption is good to go! PTAL folks

File: mkdocs.yml
Comment (fao89) [inline mkdocs.yml:30]: actually, we should remove this file, mkdocs isn't used anymore

**Review by fao89 — APPROVED** [review]

File: mkdocs.yml
Comment (jistr) [reply mkdocs.yml:30]: +1, not a blocker for this patch

**Review by jistr — APPROVED** [review]
+1 but leaving space for @gibizer and @SeanMooney to review.

File: docs_user/modules/openstack-edpm_adoption.adoc
Comment (gibizer) [inline docs_user/modules/openstack-edpm_adoption.adoc:436]: nova_cell1 is in openstack-cell1-galera

File: docs_user/modules/openstack-edpm_adoption.adoc
Comment (gibizer) [inline docs_user/modules/openstack-edpm_adoption.adoc:543]: nit: s/con/can/

**Review by gibizer — APPROVED** [review]
my nits can be fixed in a follow up

File: docs_user/modules/openstack-edpm_adoption.adoc
Comment (bogdando) [reply docs_user/modules/openstack-edpm_adoption.adoc:436]: goot catch, I'll fix in follow-up

File: docs_user/modules/openstack-pull_openstack_configuration.adoc
Comment (matbu) [reply docs_user/modules/openstack-pull_openstack_configuration.adoc:94]: @jistr ack ok, no problem from my side.

File: docs_user/modules/openstack-pull_openstack_configuration.adoc
Comment (bogdando) [reply docs_user/modules/openstack-pull_openstack_configuration.adoc:97]: that's the only way to carry on those after we shut down the source cloud services...
we could use env files, like I [suggested earlier](https://github.com/openstack-k8s-operators/data-plane-adoption/pull/146), which should also help users to interrupt and resume going through the documented steps w/o the need to repeat all the "lost" exports and set commands

Comment (bogdando) [reply docs_user/modules/openstack-pull_openstack_configuration.adoc:97]: actually, that's not a simple task, to store those evaluated multi-line values in an env file so that those could be sources w/o problems... I'll try and let you know

Comment (jistr) [reply docs_user/modules/openstack-pull_openstack_configuration.adoc:97]: @bogdando I think it's ok to keep the env vars. Writing to a file would be good for persistence, but it could further inflate the docs, so it may not be a silver bullet. (?)

Comment (bogdando) [reply docs_user/modules/openstack-pull_openstack_configuration.adoc:97]: please see https://github.com/openstack-k8s-operators/data-plane-adoption/pull/271

File: docs_user/modules/openstack-edpm_adoption.adoc
Comment (bogdando) [reply docs_user/modules/openstack-edpm_adoption.adoc:543]: https://github.com/openstack-k8s-operators/data-plane-adoption/pull/272

File: mkdocs.yml
Comment (bogdando) [reply mkdocs.yml:30]: https://github.com/openstack-k8s-operators/data-plane-adoption/pull/272

**Pull request merged**
